backend/src/routes/user.ts

```diff
@@ -0,0 +1,179 @@
+        return;
+    }
+
+    res.send(await user.save());
+}
+
+async function getAllUsers(req: Request, res: Response): Promise<void> {
+    res.send(await User.find({}));
+}
+
+async function getById(req: Request, res: Response) {
+    if (!Types.ObjectId.isValid(req.params.id)) {
+        res.status(404).send({ message: `User with id: ${req.params.id} not found` });
+        return;
+    }
+
+    const result = await User.findById(req.params.id);
+
+    if (!result) {
+        res.status(404).send({ message: `User with id: ${req.params.id} not found` });
+        return;
+    }
+
+    result.exerciseEvents.map(async(elem) => {
+        return await ExerciseEvent.findById(elem);
+    });
+
+    res.send({
+        _id: result._id,
+        username: result.username,
+        profile_image: result.profile_image,
+        bio: result.bio,
+        images: result.images,
+        exerciseEvents: result.exerciseEvents,
+    });
+}
+
+async function loginUser(req: Request, res: Response) {
+    let username = null;
+    let result = null;
+
+    try {
+        username = req.body.username;
+
+    } catch (err) {
+        console.error(err);
+        res.status(404).send({ message: `User with username: ${username} not found` });
+        return;
+    }
+
+    result = await User.findOne({ username: username });
+
+    if (!result) {
+        res.status(404).send({ message: `User with username: ${username} not found` });
+        return;
+    }
+
+    try {
+        let password = req.body.password;
+
+        if (password !== result.password) {
+            res.status(404).send({ message: `Invalid password` });
+            return;
+        }
+
+        res.send({
+            _id: result._id,
+            username: result.username,
+        });
+    } catch (err) {
+        console.error(err);
+        res.status(404).send({ message: `Invalid password` });
+        return;
+    }
+}
+
+async function updateUser(req: Request, res: Response) {
+    let id = req.params.id;
+
+    if (!Types.ObjectId.isValid(id)) {
+        res.status(404).send({ message: `User with id: ${id} not found` });
+        return;
+    }
+
+    let user = await User.findById(id);
+
+    if (!user) {
+        res.status(404).send({ message: `User with id ${id} not found` });
+    }
+
+    let err = false;
+    let toUpdate = { 
+        username: req.body.username || user.username,
+        password: req.body.password || user.password,
+        profile_image: req.body.profile_image || user.profile_image,
+        bio: req.body.bio || user.bio,
+        images: user.images,
+        exerciseEvents: user.exerciseEvents,
+
+    };
+
+    if (req.body.images) {
+        req.body.images.forEach(async(elem) => {
+            if (toUpdate.images.indexOf(elem) === -1) {
+                toUpdate.images.push(elem);
+            }
+        });
+    }
+
+    if (req.body.exerciseEvents) {
+        for (let x in req.body.exerciseEvents) {
```

Comment:
why not use map?

Reply:
annoying issues with scoping/detecting if there was an error parsing the eventId - not really sure if it would be better to set a variable or something when there's an error to ensure we don't send 2 responses to the client